lib/ddtrace/logger.rb

```diff
@@ -10,30 +11,89 @@ class Logger < ::Logger
     def initialize(*args, &block)
       super
       self.progname = LOG_PREFIX
+
+      # Rate limit buckets
+      @buckets = {}
     end
 
     def add(severity, message = nil, progname = nil, &block)
       where = ''
+      c = nil
+      skipped_messages = ''
+
+      # Logging rate of 0 means log everything, so skip bucket checking
+      if Datadog.configuration.logging_rate > 0
+        # Get the call stack to determine who is logging
+        c = caller
+
+        # Get the rate limit bucket key
+        key = rate_limit_key(severity, progname, c)
+
+        # Get the currently existing bucket, if one exists
+        last_bucket = @buckets[key]
+
+        # Get the current time bucket
+        # DEV: This is used to determine which logging_rate period this log occurred in
+        #   e.g. (1557944138 / 60) = 25965735
+        current_time_bucket = Time.now.to_i / Datadog.configuration.logging_rate
+
+        # If no previous bucket exists or the time bucket has changed, then we can log
+        if last_bucket.nil? || last_bucket[:time_bucket] != current_time_bucket
+          # Get the previous number of skipped messages to append to message
+          unless last_bucket.nil? || last_bucket[:skipped].zero?
+            skipped_messages = ", #{last_bucket[:skipped]} additional messages skipped"
+          end
+
+          # We are in a new time bucket, and are not rate limited
+          # Reset the bucket for this key
+          @buckets[key] = { time_bucket: current_time_bucket,
+                            skipped: 0 }
+
+        # Otherwise, we are in an existing bucket we already logged for
+        else
+          # Increment the skipped count and return without logging anything
+          @buckets[key][:skipped] += 1
```

Comment:
What would it take to create an integer overflow in a single bucket? How fast would they have to log to cause this?

Reply:
oh wow, that would... take a lot.

They'd have to log `Interger::MAX` times in 60 seconds.

Even then an overflow here wouldn't effect the code at all, this will still work as expected, all that will happen is the log message saying how many were skipped will show a negative number

Reply:
Fair enough, just wanted to be sure it doesn't create some kind of error.